src/wh_client_cryptocb.c

```diff
@@ -643,103 +709,113 @@ int wh_Client_CryptoCb(int devId, wc_CryptoInfo* info, void* inCtx)
+            cmac->type = type;
+            ret = 0;
+            break;
+        }
+
+#ifdef DEBUG_CRYPTOCB_VERBOSE
+        printf("[client] cmac key:%p key_len:%d in:%p in_len:%d out:%p out_len:%d keyId:%x\n",
+                key, key_len, in, in_len, mac, mac_len, key_id);
+#endif
+        uint16_t group = WH_MESSAGE_GROUP_CRYPTO;
+        uint16_t action = WC_ALGO_TYPE_CMAC;
+
+
+        wh_Packet_cmac_req* req = &packet->cmacReq;
+        uint8_t* req_in = (uint8_t*)(req + 1);
+        uint8_t* req_key = req_in + in_len;
+        uint16_t req_len = WH_PACKET_STUB_SIZE + sizeof(*req) +
+                in_len + key_len;
+
+        if (req_len > WOLFHSM_CFG_COMM_DATA_LEN) {
+            /* if we're using an HSM req_key return BAD_FUNC_ARG */
+            if (!WH_KEYID_ISERASED(key_id)) {
                 ret = BAD_FUNC_ARG;
             } else {
                 ret = CRYPTOCB_UNAVAILABLE;
             }
             break;
         }
-        /* Return success for init call with NULL params */
-        if (    (info->cmac.in == NULL) &&
-                (info->cmac.key == NULL) &&
-                (info->cmac.out == NULL)) {
-            ret = 0;
-            break;
-        }
 
-        packet->cmacReq.type = info->cmac.type;
-        packet->cmacReq.keyId = (intptr_t)info->cmac.cmac->devCtx;
+        memset(req, 0 , sizeof(*req));
+        req->type = type;
+        req->keyId = key_id;
         /* multiple modes are possible so we need to set zero size if buffers
          * are NULL */
-        if (info->cmac.in != NULL) {
-            packet->cmacReq.inSz = info->cmac.inSz;
-            XMEMCPY(in, info->cmac.in, info->cmac.inSz);
+        req->inSz = in_len;
+        if (in_len != 0) {
+            memcpy(req_in, in, in_len);
         }
-        else
-            packet->cmacReq.inSz = 0;
-        if (info->cmac.key != NULL) {
-            packet->cmacReq.keySz = info->cmac.keySz;
-            XMEMCPY(key, info->cmac.key, info->cmac.keySz);
+        req->keySz = key_len;
+        if (key_len != 0) {
+            memcpy(req_key, key, key_len);
         }
-        else
-            packet->cmacReq.keySz = 0;
-        if (info->cmac.out != NULL)
-            packet->cmacReq.outSz = *(info->cmac.outSz);
-        else
-            packet->cmacReq.outSz = 0;
+        req->outSz = mac_len;
+
         /* write request */
-        ret = wh_Client_SendRequest(ctx, group, WC_ALGO_TYPE_CMAC,
-            WH_PACKET_STUB_SIZE + sizeof(packet->cmacReq) +
-            packet->cmacReq.inSz + packet->cmacReq.keySz, (uint8_t*)packet);
+        ret = wh_Client_SendRequest(ctx, group, action, req_len,
+                (uint8_t*)packet);
         if (ret == 0) {
+            /* Update the local type since call succeeded */
+            cmac->type = type;
             /* if the client marked they may want to cancel, handle the
-             * response in a seperate call */
+             * response req_in a seperate call */
```

Comment:
```suggestion
             * response req_in a separate call */
```

Also, `req_in` is the request input data, what were you trying for with this edit?

Reply:
Used refactor function in eclipse editor.  It found some comments when i was renaming "in" to "req_in". ill fix.